### PR TITLE
65 implement authentication and authorization on certain routes

### LINKED ISSUE #65: Implement authentication and authorization on certain routes
https://stackoverflow.blog/2021/10/06/best-practices-for-authentication-and-authorization-for-rest-apis/
https://developer.mozilla.org/fr/docs/Web/HTTP/Headers/Authorization

We will be using header `Authorization Basic id:password` where `id:password` is in base64. In addition, we will use a middleware in gin to verify credentials. Continue to the route if it is okay and send 401 if not.
Need to implement WWW-Authenticate as well https://developer.mozilla.org/fr/docs/Web/HTTP/Headers/WWW-Authenticate

### PR DESCRIPTION
**Pull request by Brawdunoir**
This PR add an authentication middleware. See #65.
This middleware is implemented in all routes except the `/users` on `POST`, because it is our equivalent to /login.

This middleware uses basic auth and returns 401 code if
- `Authorization` header is empty or not well formed
- The password mismatch from the one in DB

It returns 404 code if no user with this id is found in DB

It also sets the found and authenticated user as a context value so it can be retrieved in different routes.
We use this value on `/users/:id` for `PATCH` and `DELETE` requests to assert the request author is the same as the targeted user. If not, we also send a 401 code.

This PR has a lot of changements in tests utils to implement the header and keep the tests running.

### CODE REVIEW

File: routes/user.go [hunk 1]
```diff
@@ -29,6 +29,11 @@ func CreateUser(c *gin.Context) {
 		return
 	}
 
+	// Create 64 bytes random password
+	password := make([]byte, 32)
```
Comment (WhiteDiamondz) [inline routes/user.go:33]: 32 ?

File: tests/middleware_test.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+	method := http.MethodPost
+	tests := utils.TestCreate{
+		Target:  roomURL,
+		Headers: headers,
+		SubTests: []utils.SubTest{
+			{Name: "Wrong password", Request: utils.Request{Method: method, Headers: utils.GetBasicAuthHeader(id, "password")}, ResponseCode: http.StatusUnauthorized, ResponseBodyRegex: `{"error":"User not authorized"}`},
+			{Name: "User not found", Request: utils.Request{Method: method, Headers: utils.GetBasicAuthHeader("987654321", "password")}, ResponseCode: http.StatusNotFound, ResponseBodyRegex: `{"error":"User not found"}`},
+			{Name: "Empty authorization header", Request: utils.Request{Method: method, Headers: []utils.Header{}}, ResponseCode: http.StatusUnauthorized, ResponseBodyRegex: `{"error":"User not authorized"}`},
+			{Name: "Not well formed authorization header", Request: utils.Request{Method: method, Headers: []utils.Header{{Key: "Authorization", Value: "apikey xxx"}}}, ResponseCode: http.StatusUnauthorized, ResponseBodyRegex: `{"error":"User not authorized"}`},
```
Comment (WhiteDiamondz) [inline tests/middleware_test.go]:
```suggestion
			{Name: "Badly formed authorization header", Request: utils.Request{Method: method, Headers: []utils.Header{{Key: "Authorization", Value: "apikey xxx"}}}, ResponseCode: http.StatusUnauthorized, ResponseBodyRegex: `{"error":"User not authorized"}`},
```

File: utils/routes/conversion.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+func ExtractUserFromContext(c *gin.Context) (*models.User, error) {
+	val, ok := c.Get(variables.USER_CONTEXT_KEY)
+	if !ok {
+		return nil, errors.New("user not found in context")
+	}
+	user, ok := val.(models.User)
+	if !ok {
+		return nil, errors.New("value found in context but it was not a user")
```
Comment (WhiteDiamondz) [inline utils/routes/conversion.go:18]: In what case could this happen?

File: routes/user.go (hunk 1 above)
Comment (Brawdunoir) [reply routes/user.go:33]: Indeed, it is fixed

File: routes/user.go [hunk 4]
```diff
@@ -29,6 +29,11 @@ func CreateUser(c *gin.Context) {
 		return
 	}
 
+	// Create 64 bytes random password
```
Comment (WhiteDiamondz) [inline routes/user.go]: Change to 32 then ?

File: utils/tests/tests.go [hunk 5]
```diff
@@ -23,21 +28,69 @@ type ITest interface {
...
+	// ResponseBodyRegex is the expected response body regex.
 	ResponseBodyRegex interface{}
 }
 
-// Request a simple request to be sent to the router.
+// Request is a simple request to be sent to the router.
 type Request struct {
+	// Method is the HTTP method to use.
 	Method string
+	// Target is append to the Test Target URL.
```
Comment (WhiteDiamondz) [inline utils/tests/tests.go]:
```suggestion
	// Target is appended to the Test Target URL.
```

**Review by WhiteDiamondz — CHANGES_REQUESTED** [review]
Small changes, otherwise almost ready to go through !

File: utils/routes/conversion.go (hunk 3 above)
Comment (Brawdunoir) [reply utils/routes/conversion.go:18]: None, except if in middleware we don’t set correctly the value of key user.
Basically it should never happen ;)

File: routes/user.go (hunk 4 above)
Comment (Brawdunoir) [reply routes/user.go]: Has been done.

**Review by WhiteDiamondz — APPROVED** [review]

**Pull request merged**